### PR TITLE
[SPARK-27909][SQL] Do not run analysis inside CTE substitution

### PR DESCRIPTION
**Pull request by rdblue**
## What changes were proposed in this pull request?

This updates CTE substitution to avoid needing to run all resolution rules on each substituted expression. Running resolution rules was previously used to avoid infinite recursion. In the updated rule, CTE plans are substituted as sub-queries from right to left. Using this scope-based order, it is not necessary to replace multiple CTEs at the same time using `resolveOperatorsDown`. Instead, `resolveOperatorsUp` is used to replace each CTE individually.

By resolving using `resolveOperatorsUp`, this no longer needs to run all analyzer rules on each substituted expression. Previously, this was done to apply `ResolveRelations`, which would throw an `AnalysisException` for all unresolved relations so that unresolved relations that may cause recursive substitutions were not left in the plan. Because this is no longer needed, `ResolveRelations` no longer needs to throw `AnalysisException` and resolution can be done in multiple rules.

## How was this patch tested?

Existing tests in `SQLQueryTestSuite`, `cte.sql`.

### CODE REVIEW

**Comment by rdblue** [comment]
@cloud-fan, can you take a look at this?

These changes are needed to get the DSv2 table resolution fixed because that adds a second rule to resolve relations. That means that `ResolveRelations` should not throw `AnalysisException` if it doesn't find a table because another resolution rule may be able to find the table.

**Comment by rdblue** [comment]
@jzhuge, FYI

**Comment by cloud-fan** [comment]
looks reasonable to me, but not very familiar with this part, cc @gatorsmile @hvanhovell

**Comment by jzhuge** [comment]
LGTM. This PR not only fixes cte.sql failures we encountered while enhancing #24741, it also makes the code much easier to reason about.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 1]
```diff
@@ -215,23 +215,26 @@ class Analyzer(
   object CTESubstitution extends Rule[LogicalPlan] {
     def apply(plan: LogicalPlan): LogicalPlan = plan.resolveOperatorsUp {
       case With(child, relations) =>
-        substituteCTE(child, relations.foldLeft(Seq.empty[(String, LogicalPlan)]) {
-          case (resolved, (name, relation)) =>
-            resolved :+ name -> executeSameContext(substituteCTE(relation, resolved))
-        })
+        // substitute CTE expressions right-to-left to resolve references to previous CTEs:
+        // with a as (select * from t) with b as (select * from a) select * from b
+        relations.reverse.foldLeft(child) {
```
Comment (hvanhovell) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: OCD: is `relations.reverse.foldLeft(..)` not similar to `relations.foldRight(..)`?

**Review by hvanhovell — APPROVED** [review]
LGTM

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala (hunk 1 above)
Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: good idea about `foldRight`, it saves the creation of an intermedia collection.

**Comment by dongjoon-hyun** [comment]
I agree with the above comments on `foldRight` and the rest of PR also looks good to me.

**Review by gatorsmile — COMMENTED** [review]
The JIRA https://issues.apache.org/jira/browse/SPARK-27909 said it is a bug fix. I think we already fixed the infinite recursion in https://github.com/apache/spark/pull/14397. 

Is this PR just a refactoring?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 2]
```diff
@@ -215,23 +215,26 @@ class Analyzer(
   object CTESubstitution extends Rule[LogicalPlan] {
     def apply(plan: LogicalPlan): LogicalPlan = plan.resolveOperatorsUp {
       case With(child, relations) =>
-        substituteCTE(child, relations.foldLeft(Seq.empty[(String, LogicalPlan)]) {
-          case (resolved, (name, relation)) =>
-            resolved :+ name -> executeSameContext(substituteCTE(relation, resolved))
-        })
+        // substitute CTE expressions right-to-left to resolve references to previous CTEs:
+        // with a as (select * from t) with b as (select * from a) select * from b
```
Comment (peter-toth) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: Just a minor note that this example should read `with a as (select * from t), b as (select * from a) select * from b` as double `with` aren't allowed.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala (hunk 1 above)
Comment (rdblue) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: Fixed.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala (hunk 2 above)
Comment (gatorsmile) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: Yes. This example needs an update.

Comment (rdblue) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: Done.

**Comment by rdblue** [comment]
@dongjoon-hyun, I think we have addressed all review comments, could you have another look? Thank you!

**Review by jzhuge — APPROVED** [review]

**Comment by gatorsmile** [comment]
LGTM

Merged to master.

**Comment by rdblue** [comment]
Thanks for merging this, @gatorsmile!

**Comment by dongjoon-hyun** [comment]
Late LGTM! Thank you for merging, too.

**Pull request closed without merging**